File: 31_oct_level_2/problem_17_top_k_frequent_elements/solution.py
from typing import TextIO, List
import sys

class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        from collections import Counter
        import heapq
        
        # Count frequencies
        freq = Counter(nums)
        
        # Use heap to get top k
        # Python's heapq is min heap, so negate frequencies for max heap behavior
        result = heapq.nlargest(k, freq.keys(), key=freq.get)
        
        # Sort result for consistent output
        result.sort()
        return result

def solve(infile: TextIO = sys.stdin, outfile: TextIO = sys.stdout) -> None:
    n, k = map(int, infile.readline().split())
    nums = list(map(int, infile.readline().split()))
    
    solution = Solution()
    result = solution.topKFrequent(nums, k)
    
    outfile.write(' '.join(map(str, result)) + '\n')

if __name__ == "__main__":
    solve()

File: 31_oct_level_2/problem_17_top_k_frequent_elements/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Count frequencies
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }
        
        // Use min heap to keep top k elements
        auto cmp = [](pair<int, int>& a, pair<int, int>& b) {
            return a.second > b.second;
        };
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> minHeap(cmp);
        
        for (auto& p : freq) {
            minHeap.push(p);
            if (minHeap.size() > k) {
                minHeap.pop();
            }
        }
        
        // Extract results
        vector<int> result;
        while (!minHeap.empty()) {
            result.push_back(minHeap.top().first);
            minHeap.pop();
        }
        
        // Sort result for consistent output
        sort(result.begin(), result.end());
        
        return result;
    }
};

void solve() {
    int n, k;
    cin >> n >> k;
    
    vector<int> nums(n);
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }
    
    Solution solution;
    vector<int> result = solution.topKFrequent(nums, k);
    
    for (int i = 0; i < result.size(); i++) {
        if (i > 0) cout << " ";
        cout << result[i];
    }
    cout << endl;
}

int main() {
    solve();
    return 0;
}
